### PR TITLE
[mono][jit] Fix a JIT assert.

### LINKED ISSUE #57560: [wasm] `mono-aot-cross` asserts with `function:mono_method_to_ir, VAR 1 (TValue_REF) cannot be expanded in this context with 1 instantiations `
### Description

Migrated a Blazor application from .NET 5 to latest 6 preview and wanted to test AOT which led me to this error.
Managed to narrow it down to a dictionary implementation which had a strange equality operator overload using `EqualityComparer<>.Default.Equals(left, right);`. Changing it to just `Equals(left, right);` fixed it (`IEquatable<>` is implemented). I have a full repro in the zip-file below.

```pwsh
> dotnet publish -c Release
Microsoft (R) Build Engine version 17.0.0-preview-21378-03+d592862ed for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  SomeTestLib -> C:\code\private\wasmtest\SomeTestLib\bin\Release\net6.0\SomeTestLib.dll
  wasmtest -> C:\code\private\wasmtest\wasmtest\bin\Release\net6.0\wasmtest.dll
  wasmtest (Blazor output) -> C:\code\private\wasmtest\wasmtest\bin\Release\net6.0\wwwroot
  Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
  AOT'ing 30 assemblies
C:\Program Files\dotnet\packs\Microsoft.NET.Runtime.WebAssembly.Sdk\6.0.0-preview.7.21377.19\Sdk\WasmApp.Native.targets(443,5): error : Precompiling failed for C:\code\private\wasmtest\wasmtest\obj\Release\net6.0\linked\SomeTestLib.dll: Mono Ahead of Time compiler - compiling assembly C:\code\private\wasmtest\wasmtest\obj\Release\net6.0\linked\SomeTestLib.dll [C:\code\private\wasmtest\wasmtest\wasmtest.csproj]
C:\Program Files\dotnet\packs\Microsoft.NET.Runtime.WebAssembly.Sdk\6.0.0-preview.7.21377.19\Sdk\WasmApp.Native.targets(443,5): error : * Assertion at D:\workspace\_work\1\s\src\mono\mono\mini\method-to-ir.c:7318, condition `is_ok (error)' not met, function:mono_method_to_ir, VAR 1 (TValue_REF) cannot be expanded in this context with 1 instantiations [C:\code\private\wasmtest\wasmtest\wasmtest.csproj]
```

Repro solution:
[wasmtest.zip](https://github.com/dotnet/runtime/files/6998976/wasmtest.zip)

In `Classes.cs`, make the following modification to let the build complete successfully with AOT.

```diff
    public static bool operator ==(ValueDictionary<TKey, TValue>? left, ValueDictionary<TKey, TValue>? right) =>
-        EqualityComparer<ValueDictionary<TKey, TValue>>.Default.Equals(left, right);
+        Equals(left, right);
```

`dotnet publish -c Release`

### Configuration

* Windows 10, x64
```pwsh
PS > dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.100-preview.7.21379.14
 Commit:    22d70b47bc

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19043
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\

Host (useful for support):
  Version: 6.0.0-preview.7.21377.19
  Commit:  91ba01788d
```

### Regression?

I haven't tried to reproduce in older previews.

### PR DESCRIPTION
**Pull request by vargaz**
Fixes https://github.com/dotnet/runtime/issues/57560.

### CODE REVIEW

File: src/mono/mono/mini/method-to-ir.c [hunk 1]
```diff
@@ -7336,9 +7336,10 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 				ERROR_DECL (error);
 
 				MonoMethod *new_cmethod = mono_class_get_virtual_method (sp [0]->klass, cmethod, error);
-				mono_error_assert_ok (error);
-				cmethod = new_cmethod;
-				virtual_ = FALSE;
+				if (is_ok (error)) {
+					cmethod = new_cmethod;
+					virtual_ = FALSE;
+				}
```
Comment (lambdageek) [inline src/mono/mono/mini/method-to-ir.c:7342]:
```suggestion
				} else
					mono_error_cleanup (error);
```

**Review by lambdageek — APPROVED** [review]

**Comment by radical** [comment]
Can we add a test for this?

**Review by lewing — APPROVED** [review]

**Pull request merged**